Add Product Discovery section to homepage [MAILPOET-4827]

costasovo opened the pull request:
## Description
This PR adds the product discovery section to the homepage
<img width="757" alt="Screenshot 2023-01-12 at 16 32 19" src="https://user-images.githubusercontent.com/1082140/212110031-e245a623-a816-4d6a-bfe9-be30a16e2e49.png">


## Code review notes
I recommend reviewing commit by commit.

## QA notes
When testing using the ZIP build, the illustration images in the task list won't load because they will be publicly available after the release.
You can check the feature with the images by testing with the dev env and adding `define( 'MAILPOET_USE_CDN', false ); ` to the wp-config.php. Alternatively, you can ping me and I'll push the images to the repo so that they also work from the zip.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-4827]

## After-merge notes

_N/A_


[MAILPOET-4827]: https://mailpoet.atlassian.net/browse/MAILPOET-4827?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

costasovo committed:
Add basic product discovery component

[MAILPOET-4827]

costasovo committed:
Add functionality for hiding the product discovery section

[MAILPOET-4827]

costasovo committed:
Add simple product discovery tasks

[MAILPOET-4827]

costasovo committed:
Add illustration assets for product discovery tasks

[MAILPOET-4827]

costasovo committed:
Display illustrations in product discovery tasks

[MAILPOET-4827]

costasovo committed:
Refactor homepage contents to an extra component

[MAILPOET-4827]

costasovo committed:
Refactor section hiding from specific homepage sections to parent component

This was done so that we can safely skip loading data for hidden sections.
[MAILPOET-4827]

costasovo committed:
Refactor homepage data fetching to a new service

[MAILPOET-4827]

costasovo committed:
Add test for HomepageDataController

[MAILPOET-4827]

costasovo committed:
Fetch task list status only when task list is displayed

[MAILPOET-4827]

costasovo committed:
Add product discovery status detection to homepage data service

[MAILPOET-4827]

costasovo committed:
Display product discovery tasks status in UI

[MAILPOET-4827]

costasovo committed:
Display different message when discovery task is done

[MAILPOET-4827]

costasovo committed:
Display proper product discovery tasks if Woo is active

[MAILPOET-4827]

costasovo committed:
Style product discovery section

[MAILPOET-4827]

costasovo committed:
Add css for small screens for product discovery

[MAILPOET-4827]

costasovo committed:
Add dismiss product discovery CTA when all tasks are done

[MAILPOET-4827]

costasovo committed:
Improve mixpanel track wrapper to support options and callback params

When tracking links we need to use callback parameter to ensure the event is tracked
before the page is redirected to the link.
See https://developer.mixpanel.com/docs/javascript-full-api-reference#mixpaneltrack
[MAILPOET-4827]

costasovo committed:
Add tracking of homepage discovery tasks clicks

[MAILPOET-4827]

costasovo committed:
Add check from automation flows with welcome emails

This commit improves the check that a user has set a welcome email campaign (a task in product discovery task list)
by checking also for automation flows that are triggered by new subscriber or user and contain
send email action.
[MAILPOET-4827]

costasovo committed:
Make complete tasks in product discovery non clickable

[MAILPOET-4827]

costasovo committed:
Extend homepage acceptance test to check also product discovery section

[MAILPOET-4827]

costasovo committed:
Refactor always use camel case within homepage data passed to JS

[MAILPOET-4827]

lysyjan commented on mailpoet/assets/js/src/homepage/components/product-discovery.tsx:
```diff
@@ -87,8 +89,9 @@ export function ProductDiscovery({ onHide }: Props): JSX.Element {
-    </div>
+      }
+    >
+      <ul className="mceProgress">{tasks.map((item) => item)}</ul>
```
The class `mceProgress` looks weird to me. I haven't noticed it in another commit and haven't found it in our codebase.

costasovo replied:
Ha! Nice catch!  I'll remove it. I wonder where it came from. The copilot maybe?

lysyjan reviewed (commented):
Great work! I found just one little thing I would like to ask.

costasovo committed:
Remove unused classname from product discovery list

[MAILPOET-4827]

costasovo committed:
Add semantic markup for homepage section component

[MAILPOET-4827]

costasovo commented:
@lysyjan, I rebased the PR on the latest trunk, removed the strange class name, and added one more commit where I slightly improved the homepage section component. I forgot to do it in the original PR.

Aschepikov commented:
Hi @costasovo  there is some error, can you take a look pls
![image](https://user-images.githubusercontent.com/113185437/214050839-297f09dd-b822-4296-8971-b215989f1a9d.png)
`TypeError: Cannot destructure property 'isTaskListHidden' of '(0 , k.Z)(...)' as it is undefined. 
File name: No fileName reported 
Stack trace: TypeError: Cannot destructure property 'isTaskListHidden' of '(0 , k.Z)(...)' as it is undefined.
    at z (https://brown-parrot.jurassic.ninja/wp-content/plugins/mailpoet/assets/dist/js/admin.js?ver=4.4.0:2:8766)
    at ra (https://brown-parrot.jurassic.ninja/wp-content/plugins/mailpoet/assets/dist/js/commons.js?ver=4.4.0:9:1396148)
    at js (https://brown-parrot.jurassic.ninja/wp-content/plugins/mailpoet/assets/dist/js/commons.js?ver=4.4.0:9:1448197)
    at xl (https://brown-parrot.jurassic.ninja/wp-content/plugins/mailpoet/assets/dist/js/commons.js?ver=4.4.0:9:1435471)
    at bl (https://brown-parrot.jurassic.ninja/wp-content/plugins/mailpoet/assets/dist/js/commons.js?ver=4.4.0:9:1435399)
    at wl (https://brown-parrot.jurassic.ninja/wp-content/plugins/mailpoet/assets/dist/js/commons.js?ver=4.4.0:9:1435262)
    at fl (https://brown-parrot.jurassic.ninja/wp-content/plugins/mailpoet/assets/dist/js/commons.js?ver=4.4.0:9:1432249)
    at https://brown-parrot.jurassic.ninja/wp-content/plugins/mailpoet/assets/dist/js/commons.js?ver=4.4.0:9:1382070
    at t.unstable_runWithPriority (https://brown-parrot.jurassic.ninja/wp-content/plugins/mailpoet/assets/dist/js/commons.js?ver=4.4.0:9:1585646)
    at ji (https://brown-parrot.jurassic.ninja/wp-content/plugins/mailpoet/assets/dist/js/commons.js?ver=4.4.0:9:1381847)`

costasovo commented:
@Aschepikov I think this was caused by cached JS assets.  I changed the naming of some variables ([see commit](https://github.com/mailpoet/mailpoet/pull/4649/commits/3d7e7bd8bcda26a87409b4cfba40834c749583ab)) and if old admin.js file is loaded this is what happens. This caching issue may happen on feature branches because we use the plugin version as a versioning parameter in the asset URL.

Aschepikov reviewed (approved):
GJ!
